File: executor/bridge.js
'use strict';
/*
 * executor/bridge.js — the executor's half of the local bridge to MacroQuest.
 *
 * The executor and the MQ Lua bridge run on the SAME machine (the 32 GB box), so they
 * talk over two append-only line files in a shared bridge directory — no socket library
 * needed inside MQ's Lua, and atomic by virtue of being newline-delimited appends.
 *
 *   queue.tsv : executor APPENDS   "<id>\t<command>\n"            (mqbridge.lua consumes)
 *   acks.tsv  : mqbridge.lua APPENDS "<id>\t<status>\t<message>\n" (executor consumes)
 *
 * status is "OK" or "ERR". This module hides the file plumbing behind send()/poll().
 */

const fs = require('fs');
const path = require('path');

class Bridge {
  /** @param {string} dir bridge directory shared with mqbridge.lua */
  constructor(dir) {
    this.dir = dir;
    this.queuePath = path.join(dir, 'queue.tsv');
    this.acksPath = path.join(dir, 'acks.tsv');
    this._ackOffset = 0;            // byte offset already consumed from acks.tsv
    this._pending = new Map();      // id -> { resolve, reject, timer }
    fs.mkdirSync(dir, { recursive: true });
    // Touch the files so reads don't race file creation.
    fs.closeSync(fs.openSync(this.queuePath, 'a'));
    fs.closeSync(fs.openSync(this.acksPath, 'a'));
    // Start reading acks AFTER whatever is already there (don't replay history).
    this._ackOffset = fs.statSync(this.acksPath).size;
  }

  /**
   * Enqueue one raw MQ command for the bridge to run in-game, and resolve when its ack
   * arrives. The command must not contain a tab or newline (it's one TSV field).
   * @returns {Promise<{id, status, message}>}
   */
  send(id, command, timeoutMs = 8000) {
    if (/[\t\n]/.test(command)) {
      return Promise.reject(new Error('command may not contain tab or newline'));
    }
    const line = `${id}\t${command}\n`;
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this._pending.delete(id);
        reject(new Error(`bridge timeout: no ack for ${id} within ${timeoutMs}ms (is mqbridge.lua running?)`));
      }, timeoutMs);
      this._pending.set(id, { resolve, reject, timer });
      try {
        fs.appendFileSync(this.queuePath, line);
      } catch (e) {
        clearTimeout(timer);
        this._pending.delete(id);
        reject(e);
      }
    });
  }

  /** Read any new ack lines and settle matching pending sends. Call on an interval. */
  poll() {
    let size;
    try {
      size = fs.statSync(this.acksPath).size;
    } catch {
      return;
    }
    if (size <= this._ackOffset) return;
    const fd = fs.openSync(this.acksPath, 'r');
    try {
      const len = size - this._ackOffset;
      const buf = Buffer.allocUnsafe(len);
      const read = fs.readSync(fd, buf, 0, len, this._ackOffset);
      const text = buf.toString('utf8', 0, read);
      // Only consume up to the last complete line; leave any partial tail for next poll.
      const lastNl = text.lastIndexOf('\n');
      if (lastNl === -1) return; // no complete line yet
      const complete = text.slice(0, lastNl);
      this._ackOffset += Buffer.byteLength(text.slice(0, lastNl + 1), 'utf8');
      for (const raw of complete.split('\n')) {
        if (!raw) continue;
        const [id, status, ...rest] = raw.split('\t');
        const message = rest.join('\t');
        const p = this._pending.get(id);
        if (p) {
          clearTimeout(p.timer);
          this._pending.delete(id);
          p.resolve({ id, status, message });
        }
      }
    } finally {
      fs.closeSync(fd);
    }
  }

  /** Start the ack poller. Returns a stop() function. */
  startPolling(intervalMs = 100) {
    const h = setInterval(() => this.poll(), intervalMs);
    if (h.unref) h.unref();
    return () => clearInterval(h);
  }
}

module.exports = { Bridge };

File: mqbridge/bridge.lua
--[[
  mqbridge/bridge.lua — the in-game half of the bridge (runs INSIDE MacroQuest).

  This is the THIN hands. It does no deciding and knows nothing about the contract.
  It tails a queue file the executor appends to, runs each command in-game via mq.cmd,
  and appends an ack the executor reads back. One MQ instance ("the anchor") runs this;
  it reaches the other clients via the commands the executor sends (e.g. /dex <peer> ...).

  Protocol (must match executor/bridge.js):
    queue.tsv : "<id>\t<command>\n"            (executor appends, we consume)
    acks.tsv  : "<id>\t<status>\t<message>\n"  (we append, executor consumes)
    status is "OK" (command dispatched) or "ERR" (mq.cmd raised).

  Run in-game:  /lua run bridge
  Stop:         /lua stop bridge

  IMPORTANT: set BRIDGE_DIR to the SAME directory as the executor's EQ_BRIDGE_DIR.
  Forward slashes are fine on Windows. Both processes must be on the same machine.
]]--

local mq = require('mq')

local BRIDGE_DIR = 'C:/EQ/.bridge'        -- <-- must match executor EQ_BRIDGE_DIR
local QUEUE = BRIDGE_DIR .. '/queue.tsv'
local ACKS  = BRIDGE_DIR .. '/acks.tsv'
local POLL_MS = 100

-- Byte offset already consumed from the queue file.
local offset = 0

local function fileSize(path)
  local f = io.open(path, 'rb')
  if not f then return nil end
  local size = f:seek('end')
  f:close()
  return size
end

-- Start AFTER whatever is already in the queue so a reload doesn't replay old commands.
local function initOffset()
  offset = fileSize(QUEUE) or 0
end

local function appendAck(id, status, message)
  local f = io.open(ACKS, 'ab')
  if not f then
    print(string.format('[mqbridge] ERROR: cannot open acks file %s', ACKS))
    return
  end
  -- message must not contain tab/newline; sanitize defensively.
  message = (message or ''):gsub('[\t\n]', ' ')
  f:write(string.format('%s\t%s\t%s\n', id, status, message))
  f:close()
end

local function runCommand(id, command)
  print(string.format('[mqbridge] %s -> %s', id, command))
  local ok, err = pcall(function() mq.cmd(command) end)
  if ok then
    appendAck(id, 'OK', command)
  else
    appendAck(id, 'ERR', tostring(err))
  end
end

-- Read any complete new lines from the queue and dispatch them.
local function pollQueue()
  local size = fileSize(QUEUE)
  if not size or size <= offset then return end
  local f = io.open(QUEUE, 'rb')
  if not f then return end
  f:seek('set', offset)
  local chunk = f:read('a') or ''
  f:close()
  -- Find the final newline; process only complete lines, leave any partial tail.
  local pos = nil
  for i = #chunk, 1, -1 do
    if chunk:sub(i, i) == '\n' then pos = i break end
  end
  if not pos then return end               -- no complete line yet
  local complete = chunk:sub(1, pos)
  offset = offset + #complete              -- advance by bytes consumed
  for line in complete:gmatch('[^\n]+') do
    local id, command = line:match('^([^\t]*)\t(.*)$')
    if id and command then
      runCommand(id, command)
      mq.delay(1)                          -- let the command buffer breathe
    else
      print('[mqbridge] skipping malformed line: ' .. line)
    end
  end
end

print('[mqbridge] starting. dir=' .. BRIDGE_DIR)
initOffset()
print(string.format('[mqbridge] tailing %s from offset %d', QUEUE, offset))

while true do
  pollQueue()
  mq.delay(POLL_MS)
end
